barretenberg/cpp/src/barretenberg/crypto/poseidon2/sponge/sponge.hpp

```diff
@@ -0,0 +1,167 @@
+            // If we're absorbing, and the cache is not full, add the input into the cache
+            cache[cache_size] = input;
+            cache_size += 1;
+        } else if (mode == Mode::SQUEEZE) {
+            // If we're in squeeze mode, switch to absorb mode and add the input into the cache.
+            // N.B. I don't think this code path can be reached?!
+            cache[0] = input;
+            cache_size = 1;
+            mode = Mode::ABSORB;
+        }
+    }
+
+    FF squeeze()
+    {
+        if (mode == Mode::SQUEEZE && cache_size == 0) {
+            // If we're in squeze mode and the cache is empty, there is nothing left to squeeze out of the sponge!
+            // Switch to absorb mode.
+            mode = Mode::ABSORB;
+            cache_size = 0;
+        }
+        if (mode == Mode::ABSORB) {
+            // If we're in absorb mode, apply sponge permutation to compress the cache, populate cache with compressed
+            // state and switch to squeeze mode. Note: this code block will execute if the previous `if` condition was
+            // matched
+            auto new_output_elements = perform_duplex();
+            mode = Mode::SQUEEZE;
+            for (size_t i = 0; i < rate; ++i) {
+                cache[i] = new_output_elements[i];
+            }
+            cache_size = rate;
+        }
+        // By this point, we should have a non-empty cache. Pop one item off the top of the cache and return it.
+        FF result = cache[0];
+        for (size_t i = 1; i < cache_size; ++i) {
+            cache[i - 1] = cache[i];
+        }
+        cache_size -= 1;
+        return result;
+    }
+
+    /**
+     * @brief Use the sponge to hash an input string
+     *
+     * @tparam out_len
+     * @tparam is_variable_length. Distinguishes between hashes where the preimage length is constant/not constant
+     * @param input
+     * @return std::array<FF, out_len>
+     */
+    template <size_t out_len, bool is_variable_length> static std::array<FF, out_len> hash_internal(std::span<FF> input)
+    {
+        size_t in_len = input.size();
+        const uint256_t iv = (static_cast<uint256_t>(in_len) << 64) + out_len - 1;
+        FieldSponge sponge(iv);
+
+        for (size_t i = 0; i < in_len; ++i) {
+            sponge.absorb(input[i]);
+        }
+
+        // In the case where the hash preimage is variable-length, we append `1` to the end of the input, to distinguish
+        // from fixed-length hashes. (the combination of this additional field element + the hash IV ensures
+        // fixed-length and variable-length hashes do not collide)
+        if constexpr (is_variable_length) {
+            sponge.absorb(1);
+        }
+
+        std::array<FF, out_len> output;
+        for (size_t i = 0; i < out_len; ++i) {
+            output[i] = sponge.squeeze();
+        }
+        return output;
+    }
+
+    template <size_t out_len> static std::array<FF, out_len> hash_fixed_length(std::span<FF> input)
```

Comment:
Don't see how there's a differentiation between fixed length and variable length here. It seems like we only have fixed length hashing since we only take in field elements of 254 bits.

Reply:
for now. Merkle membership hashes will likely use fixed length